### PR TITLE
[EM] Allow staging ellpack on host for GPU external memory.

### PR DESCRIPTION
**Pull request by trivialfis**
- New parameter `on_host`.
- Abstract format creation and stream creation into policy classes.

### CODE REVIEW

**Comment by trivialfis** [comment]
cc @rongou .

File: src/data/ellpack_page_source.cu [hunk 1]
```diff
@@ -1,29 +1,161 @@
...
+  bst_idx_t bound_{0};
+
+ public:
+  explicit EllpackHostCacheStreamImpl(std::shared_ptr<EllpackHostCache> cache)
+      : cache_{std::move(cache)} {}
+
+  [[nodiscard]] bst_idx_t Write(void const* ptr, bst_idx_t n_bytes) {
+    auto n = cur_ptr_ + n_bytes;
+    if (n > cache_->cache.size()) {
+      cache_->Resize(n, dh::DefaultStream());
```
Comment (rongou) [inline src/data/ellpack_page_source.cu:41]: Allocating pinned memory seems pretty expensive. I wonder if it's worth calculating the total size and allocate the cache in one shot.

Comment (trivialfis) [reply src/data/ellpack_page_source.cu:41]: That requires an additional iteration over the disk cache, which is not cheap either.

Comment (trivialfis) [reply src/data/ellpack_page_source.cu:41]: The vector class uses dynamic table internally, the amortized cost should be reasonable, we can look into the initialization cost after removing the disk cache entirely.

Comment (trivialfis) [reply src/data/ellpack_page_source.cu:41]: Alternatively, as you suggested, we can let system pagable memory to back the main cache.

Comment (rongou) [reply src/data/ellpack_page_source.cu:41]: In my test, allocating a 120GB pinned memory buffer takes 5-6 seconds. But as a first pass, this is probably fine.

**Review by rongou — APPROVED** [review]

File: src/data/ellpack_page_source.cu (hunk 1 above)
Comment (trivialfis) [reply src/data/ellpack_page_source.cu:41]: It's expensive indeed. If we extrapolate on that speed assuming it's linear, then 500 GB would be 21 seconds, probably fine for the initial implementation. We will look into malloc as the main cache.

**Pull request merged**